### DISCUSSION TITLE
Достаем данные из циклов с фильтрацией и устойчивостью к росту базы

### DISCUSSION THREAD

**Discussion by Qb1cc** [totumonline/totum-ru-issues-and-discussions#75] [Как это сделать?] [answered]
В рабочей заказной программе есть функция быстрого оприходывания товара по артикулу. Грубо говоря вбиваешь в окно артикул позиции, возвращается список заказов где есть такой артикул, отмечаем что товар пришёл...

Можно ли в тотуме реализовать подобный функционал?

> Я только начал разбираться в тотуме, и не совсем понимаю что почитать для реализации данного функционала. На сколько я понял, нужно из цикла создавать в другую таблицу товар с доп-колонкой в которой будет болтаться номер цикла, а потом уже реализовать поиск по таблице с товарами.

![image](https://user-images.githubusercontent.com/84081305/153435215-9914fd60-676c-4287-bc2a-1ac8a6d18c00.png)

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### [Видео на YouTube —>](https://youtu.be/3Txvf7uDFwk)

#### [Похожий разбор с описанием подъема данных в таблицу циклов –>](https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/31)

#### [Автозаполнение в документации —>](https://docs.totum.online/autofill-tables)

#### Коды ключевых полей показанный в видео:

**names_in_cycle:**

```
= : listUniq(list: $list)

	list: selectList(table: 's6_in_cycle'; field: 'name'; where: 'remains' != 0)
```

**h_select:**

```
= : if(condition: #name_for_search = ""; then: $#lc; else: $calc)

calc: listFilter(list: $train; key: "name" = #name_for_search)

train: listTrain(list: $list)

list: selectList(table: 's6_cycles'; field: 'data'; where: 'names_in_cycle' = #name_for_search; order: 'id' asc)
```

**h_insert:**

```
// code

= : #h_select[[uniq]]
```

```
// code action [Execute the action on changed]

= : deleteList(table: $#ntn; where: 'insert' = $list)

list: listMinus(list: $#onfv; list: $#nfv)
```

**datarow:**

```
= : $filter[0]

filter: listFilter(list: #h_select; key: "uniq" = #insert)

```

**cycle:**

```
=: $split[0]

split: strSplit(str: #datarow[uniq]; separator: "-")
```

**receive:**

```
// code action [Execute the action on changed]

a1=: if(condition: $#nfv > #ordered; then: $err)
	err: errorException(text: "Error!")

a2=: set(table: 's6_in_cycle'; cycle: #cycle; field: 'received' = $#nfv; where: 'id' = #num; log: true)

a3=: reCalculate(table: $#ntn)
```

**Comment by Qb1cc** [comment]
Огромное спасибо за видеоурок!!! Могли бы вы  этот урок, добавить в демоверсию?

**Reply by ttm-alexeypolunin** [reply]
После некоторого обсуждения мы решили, что перенос этих примеров в демо будет отнимать у нас слишком много времени.

После того, как мы выпустим .38-39 версии мы сделаем этой базе свободный доступ, но без возможности закрыть транзакцию (только на просмотр).

**Comment by Qb1cc** [comment]
Так наконец с горем пополам удалось реализовать оприходование, но появилась новая задача. 

Как сделать правильно, что бы заказная автоматически меняла статус заказа, при условии что все товары в заказе, пришли? Нужно для реализации оповещений по смс о "готовом заказе"

**Reply by ttm-alexeypolunin** [reply]
Ну вы сразу начали изучение с немного продвинутой части :)

Делаем в футере в Заказе поле (техническое, видимое только Создателю) тип Число.

Мы в нем Кодом вычисляем количество строк, в которых не все позиции приходованы:

```
=: listCount(list: $list)

list: selectList(table: $#ntn; field: 'id'; where: 'remains' != 0)
```

Теперь получается если количество таких строк было неравно 0 а стало равно 0 то нам надо поставить нужный статус — делаем этому полю код-действия по изменению:

```
// code action
=: if(condition: $#onfv != 0; condition: $#nfv = 0; then: $set)
set: set(table: $#ntn; field: 'status' = 5)
```

Теперь прорабатываем еще 2 момента: первый — это возврат статуса в рабочий в случае исправления, второй — это отсутсвие ложного срабатывания при отсутствии строк в заказе вообще:

```
// code action
a1=: if(condition: $null != $#lc; condition: $#onfv != 0; condition: $#nfv = 0; then: $set)
set: set(table: $#ntn; field: 'status' = 5)

a2=: if(condition: $null != $#lc; condition: $#onfv = 0; condition: $#nfv != 0; then: $rollb)
rollb: set(table: $#ntn; field: 'status' = 4)

~null: selectList(table: $#ntn; field: 'id')
```

Сделайте под этот вопрос отдельный топик — я запишу видео с объяснениями всяких мелочей.

**Reply by erpic-ru** [reply]
0. Создаете простую таблицу со статусами и добавляете его в расчетную таблицу в цикле

Дальше можно действовать разными способами:

1. Поле статуса размещаете в футере ([при необходимости его можно отобразить в хедере](https://docs.totum.online/fields#fields-settings-showInWebOtherPlace)) и используете код, с условием:

```
= : if(condition: $zakaz!= 0; condition: $zakaz = $prihod; then: "нужный id статуса из справочника статусов") 

zakaz: listSum (list: $1)
1:selectlist(table:$#ntn; field: 'колонка с количеством заказанных')

prihod: listSum (list: $2)
2:selectlist(table:$#ntn; field: 'колонка с количеством поступивших')
```

2. Статус можно оставить в хедере. А коды, считающие суммы, добавить в футер под каждой колонкой заказано/поступило. Поле, которое считает сумму поступивших товаров заставить устанавливать статус при достижении условия равенства заказанных и поступивших товаров через код действия при изменении и функцию `set`:

```
= : if(condition: $#nfv !=0  condition: #nfv = #"поле с суммой заказанных"; then: $update) 

update:set(table:$#ntn; field: 'status' = "нужный ID статуса")
```

Наверное, можно ещё какие-то альтернативные варианты придумать

**Reply by erpic-ru** [reply]
Упс :) долго сочинял

**Reply by ttm-alexeypolunin** [reply]
Мне нравится идея считать полное количество по суммам — это даж проще чем предыдущее значение. Я немного поправил код @erpic-ru тк при запросе из цикла в тот же цикл можно не указывать `cycle` — они добавляются интерпретатором автоматически.
